src/Qir/Runtime/lib/QIR/QubitManager.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Qubit Manager maintains mapping between user qubit objects and
+// underlying qubit identifiers (Ids). When user program allocates
+// a qubit, Qubit Manager decides whether to allocate a fresh id or
+// reuse existing id that was previously freed. When user program
+// releases a qubit, Qubit Manager tracks it as a free qubit id.
+// Decision to reuse a qubit id is influenced by restricted reuse
+// areas. When a qubit id is freed in one section of a restricted
+// reuse area, it cannot be reused in other section of the same area.
+// Borrowing of qubits is not supported and is currently implemented
+// as plain allocation.
+
+#include <iostream>
+#include <vector>
+#include <limits>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+    using QubitIdType = ::int32_t;
+
+    // The end of free lists are marked with NoneMarker value. It is used like null for pointers.
+    // This value is non-negative just like other values in the free lists.
+    constexpr QubitIdType NoneMarker = std::numeric_limits<QubitIdType>::max();
+
+    // Explicitly allocated qubits are marked with Allocated value.
+    // Qubits implicitly allocated for borrowing are "refcounted" with values [Allocated+1 .. -2].
+    // When refcount needs to be decreased to Allocated value, the qubit is automatically released.
+    constexpr QubitIdType AllocatedMarker = std::numeric_limits<QubitIdType>::max();
+
+    // Disabled qubits are marked with this value.
+    constexpr QubitIdType DisabledMarker = -1;
+
+    // We want status array to be reasonably large if initialization for it is wrong (i.e. requested size < 1).
+    constexpr QubitIdType FallbackQubitCapacity = 8;
+
+    // Indexes in the status array can potentially be in range 0 .. int.MaxValue-1.
+    // This gives maximum capacity as int.MaxValue.
+    // Index equal to int.MaxValue doesn't exist and is reserved for 'NoneMarker' - list terminator.
+    constexpr QubitIdType MaximumQubitCapacity = std::numeric_limits<QubitIdType>::max();
```

Comment:
FYI:
(I pretend for myself that I'm reading this from scratch, familiarizing myself with the Qubit Manager)

Would be nice to see some introduction about the data structures used to store the instances of `QubitIdType`. At a glance I have an impression that the instances of `QubitIdType` are to be stored in the _arrays_. If that's the case then would be nice to see the answer to the following questions
* Is it a single array or multiple arrays? If multiple, then used for the same purpose or for different purposes? (e.g. one array for available qubits, another one for allocated, the third one for blocked, borrowed, etc.);
* Can `NoneMarker` and `AllocatedMarker` occur in the same data structure? If yes then how do the algorithms distinguish between them if they have the same value?
* What is "status array"?

And these questions:
> // The end of free lists are marked with NoneMarker value

What is free list? Is it one or many? If many then why (what is each one for)? The fragment says how the end is marked, but what are the values _in_ the free list, _before the end_?

> // Explicitly allocated qubits are marked with ..

Marked where? (in which data structure?)

> // Disabled qubits are marked with ..

Marked where? (in which data structure?)

---
I have an impression that the same data type - `QubitIdType` - will be used 
* to store the values in the array elements
* and to index the array, and also to specify the capacity (or number of elements) of the array.

If that's the case then some more comments would help about why such a decision is better than the other options. 

Because in general those are two separate/independent concepts, and they can change (get refactored) regardless of each other. E.g. the capacity can be fixed to 10 (indexes fixed to the range 0..9), but the values written to the array can grow from 0 to 0xFFFF, and then reset back to 0 and grow again.

---
To summarize, the intro comment answering all those questions would help.

Reply:
Updated comments on the classes, markers and added comment to the sharedQubitStatusArray. I believe I provided enough answers now.

Reply:
Values sored are indexes. So these are not separate concepts. See QubitIdType comment.